### PR TITLE
fix(iOS): white flash on tab change when using native stack

### LINKED ISSUE #1645: White flash on tab change with @react-navigation/bottom-tabs
### Description

When I'm using `@react-navigation/bottom-tabs` with the native stack navigator, I see a white flash on tab transition in **IOS**.

If I replace `@react-navigation/native-stack` by `@react-navigation/stack`, white flash disappear.

I think it can be related to https://github.com/software-mansion/react-native-screens/issues/1276 but I'm not sure we talk about the same issue.

https://user-images.githubusercontent.com/10655690/203101057-2af9f31c-674b-47eb-9dba-dded56bb9753.mp4

### Steps to reproduce

1. Get Minimal reproduction repo https://github.com/gkueny/native-stack-flash-issue-reproduction
2. Install dependencies
3. Run APP
4. Navigate between tab screen

### Snack or a link to a repository

https://github.com/gkueny/native-stack-flash-issue-reproduction

### Screens version

3.20.0 & 3.29.0

### React Native version

0.69.7 & 0.73.6

### Platforms

iOS

### JavaScript runtime

_No response_

### Workflow

React Native (without Expo)

### Architecture

_No response_

### Build type

Debug mode

### Device

iOS simulator

### Device model

_No response_

### Acknowledgements

Yes

### PR DESCRIPTION
**Pull request by alduzy**
## Description

This PR gets rid of undesired white flashes during `maybeAddToParentAndUpdateContainer`. The white flash was present on paper architecture when `unmountOnBlur` option was set to true on parent bottomStackNavigator (see repro).

The affected logic was previously introduced or changed by following PRs:
- https://github.com/software-mansion/react-native-screens/pull/600
- https://github.com/software-mansion/react-native-screens/pull/613
- https://github.com/software-mansion/react-native-screens/pull/643

The removed `_hasLayout` was initially added by https://github.com/software-mansion/react-native-screens/pull/600 in order to resolve an issue: https://github.com/software-mansion/react-native-screens/issues/432.
However the logic was later changed by https://github.com/software-mansion/react-native-screens/pull/613 in order to fix another issue and the added `_hasLayout` may not fix anything eventually, as stated by [this comment](https://github.com/software-mansion/react-native-screens/issues/432#issuecomment-687868469).

Fixes #1645.

## Changes

- removed `_hasLayout` variable
- added repros

### Before

https://github.com/software-mansion/react-native-screens/assets/91994767/226a32d7-728b-48dd-b21a-6a1e4195add2

### After

https://github.com/software-mansion/react-native-screens/assets/91994767/32febcf1-d159-4a9d-ae3a-373042732a6d

## Test code and steps to reproduce

- added `Test1645.js` repro to test examples
- added `Test432.tsx` repro to test examples

## Checklist

- [x] Included code example that can be used to test this change
- [x] Ensured that CI passes

### CODE REVIEW

File: ios/RNSScreenStack.mm [hunk 1]
```diff
@@ -1246,13 +1242,13 @@ - (void)prepareForRecycle
...
-  [_presentedModals removeAllObjects];
-  [_controller willMoveToParentViewController:nil];
-  [_controller removeFromParentViewController];
+  //  _invalidated = YES;
+  //  for (UIViewController *controller in _presentedModals) {
+  //    [controller dismissViewControllerAnimated:NO completion:nil];
+  //  }
+  //  [_presentedModals removeAllObjects];
+  //  [_controller willMoveToParentViewController:nil];
+  //  [_controller removeFromParentViewController];
```
Comment (tboba) [inline ios/RNSScreenStack.mm:1251]: I don't think this should be commented 😄

**Review by tboba — CHANGES_REQUESTED** [review]

File: ios/RNSScreenStack.mm (hunk 1 above)
Comment (alduzy) [reply ios/RNSScreenStack.mm]: You're totally right, I didn't notice those changes being pushed 👍

File: ios/RNSScreenStack.mm [hunk 2]
```diff
@@ -270,11 +270,7 @@ - (void)didMoveToWindow
 - (void)maybeAddToParentAndUpdateContainer
 {
   BOOL wasScreenMounted = _controller.parentViewController != nil;
-#ifdef RCT_NEW_ARCH_ENABLED
   BOOL isScreenReadyForShowing = self.window;
-#else
-  BOOL isScreenReadyForShowing = self.window && _hasLayout;
```
Comment (kkafar) [inline ios/RNSScreenStack.mm:276]: Have you checked `git blame`, why the `&& _hasLayout` was there? Don't we break any other behaviour?

**Review by kkafar — COMMENTED** [review]
I started looking and have some initial question before we can proceed

File: ios/RNSScreenStack.mm (hunk 2 above)
Comment (alduzy) [reply ios/RNSScreenStack.mm:276]: I found this: https://github.com/software-mansion/react-native-screens/pull/600 and it's related to another problem I've been investigating lately by coincidence 😄 From what I've tested it works well after removing hasLayout (both test cases).

File: ios/RNSScreenStack.mm (hunk 1 above)
Comment (alduzy) [reply ios/RNSScreenStack.mm]: https://github.com/software-mansion/react-native-screens/pull/2188/commits/35873033b1858a7bfa5cd606440c9441f6879b99

File: ios/RNSScreenStack.mm [hunk 3]
```diff
@@ -272,18 +270,10 @@ - (void)didMoveToWindow
...
-#endif
   if (!isScreenReadyForShowing && !wasScreenMounted) {
-    // We wait with adding to parent controller until the stack is mounted and has its initial
-    // layout done.
-    // If we add it before layout, some of the items (specifically items from the navigation bar),
-    // won't be able to position properly. Also the position and size of such items, even if it
-    // happens to change, won't be properly updated (this is perhaps some internal issue of UIKit).
-    // If we add it when window is not attached, some of the view transitions will be bloced (i.e.
+    // We wait with adding to parent controller until the stack is mounted
```
Comment (kkafar) [inline ios/RNSScreenStack.mm:275]:
```suggestion
    // We wait with adding to parent controller until the stack is mounted.
```

**Review by kkafar — APPROVED** [review]
I've played around with the PR for 10 minutes & haven't found any broken behaviour. 

Provided that you tested it thoroughly I believe we can proceed. 

Thanks a lot!

**Review by kkafar — APPROVED** [review]
Approving under the same conditions as above! Great job!

**Comment by alduzy** [comment]
Tested it using the provided `Test1645` and `Test432` repos, a few other `Test*` scenarios and ScreensExample app. I haven't found any undesired behaviour.

**Comment by tboba** [comment]
Let me do a quick final check on examples if this works and I guess we can proceed with this one!

**Comment by tboba** [comment]
Why are we need to test scenario from `Test432`? Is it also in the scope of this fix?

**Comment by alduzy** [comment]
@tboba The `_hasLayout` variable, that is being removed by this PR was supposed to fix #432, but It looks like it's not the case anymore. please have a look at the description for more details.

**Comment by tboba** [comment]
Alright, I understand 👍

**Comment by WoLewicki** [comment]
Did you test on all supported `iOS` versions? Maybe it was the case only on older ones?

**Comment by tboba** [comment]
I've just tested the change on iOS 15.5 and cannot spot any issues there 👍

File: ios/RNSScreenStack.mm [hunk 4]
```diff
@@ -272,18 +270,10 @@ - (void)didMoveToWindow
 - (void)maybeAddToParentAndUpdateContainer
 {
   BOOL wasScreenMounted = _controller.parentViewController != nil;
-#ifdef RCT_NEW_ARCH_ENABLED
   BOOL isScreenReadyForShowing = self.window;
```
Comment (tboba) [inline ios/RNSScreenStack.mm:273]: I don't know if keeping the definition of `isScreenReadyForShowing = self.window` is okay. I mean, by definition when the self.window is not null, this means that the screen has been mounted, so we can show it in the hierarchy. Maybe checking if self.window is not nil could be better in the `if`s below?

Comment (alduzy) [reply ios/RNSScreenStack.mm]: https://github.com/software-mansion/react-native-screens/pull/2188/commits/6a46af749d3d42369b08122f7a667ad4430aa2c0

**Review by tboba — APPROVED** [review]
LGTM!

**Comment by trickeyd** [comment]
Hey - is this fix released?

**Comment by kkafar** [comment]
Hey @trickeyd, not yet. I was planning on releasing new version after we merge this PR:

* https://github.com/software-mansion/react-native-screens/pull/2134

**Comment by kkafar** [comment]
@trickeyd, released

**Pull request merged**